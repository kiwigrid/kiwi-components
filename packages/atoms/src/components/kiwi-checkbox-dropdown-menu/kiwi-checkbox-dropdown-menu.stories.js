import { html } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/CheckboxDropdownMenu',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

export const basic = (args) => {
  const { buttonText, options } = args;

  return html`<div class="panel panel-default m-1">
    <div class="panel-body">
      <kiwi-checkbox-dropdown-menu>
        <div slot="dropdown-toggle">${buttonText}</div>
        <kiwi-checkbox-menu slot="dropdown-content">
          ${options.map(
            (o) => html`<kiwi-checkbox-menu-item>
              <kiwi-labeled-checkbox> ${o} </kiwi-labeled-checkbox>
            </kiwi-checkbox-menu-item>`,
          )}
        </kiwi-checkbox-menu>
      </kiwi-checkbox-dropdown-menu>
    </div>
  </div>`;
};
basic.args = {
  buttonText: 'Click me',
  options: ['Option 1', 'Option 2', 'Option 3'],
};

export const rightScreenEdge = (args) => {
  const { buttonText, options, menuEntryWidth } = args;

  return html`<div class="panel panel-default m-1">
    <div class="panel-body">
      <kiwi-checkbox-dropdown-menu class="pull-right">
        <div slot="dropdown-toggle">${buttonText}</div>
        <kiwi-checkbox-menu slot="dropdown-content">
          ${options.map(
            (o) => html`<kiwi-checkbox-menu-item>
              <kiwi-labeled-checkbox>
                <div style="min-width:${menuEntryWidth}">
                  ${o}
                </div></kiwi-labeled-checkbox
              >
            </kiwi-checkbox-menu-item>`,
          )}
        </kiwi-checkbox-menu>
      </kiwi-checkbox-dropdown-menu>
    </div>
  </div>`;
};
rightScreenEdge.args = {
  buttonText: 'Click me',
  options: ['Very Long Option 1', 'Very Long Option 2', 'Very Long Option 3'],
  menuEntryWidth: ['120px'],
};
