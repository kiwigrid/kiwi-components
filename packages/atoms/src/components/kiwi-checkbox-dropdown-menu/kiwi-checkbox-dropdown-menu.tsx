import {
  Component,
  ComponentInterface,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
} from '@stencil/core';

@Component({
  tag: 'kiwi-checkbox-dropdown-menu',
  shadow: false,
})
export class KiwiCheckboxDropdownMenu implements ComponentInterface {
  /** Event signaling that this dropdown was closed. */
  @Event()
  closeDropdown!: EventEmitter<void>;

  /** Type of the toggle button. */
  @Prop()
  toggleButtonType: 'default' | 'primary' | 'danger' | 'warning' | 'info' =
    'default';

  render(): JSX.Element {
    return (
      <kiwi-dropdown
        containerClass="checkbox-dropdown-menu"
        onCloseDropdown={this.handleCloseDropdown}
        toggleButtonType={this.toggleButtonType}
      >
        <slot name="dropdown-toggle" slot="dropdown-toggle" />
        <slot name="dropdown-content" slot="dropdown-content" />
      </kiwi-dropdown>
    );
  }

  private handleCloseDropdown = (): void => {
    this.closeDropdown.emit();
  };
}
