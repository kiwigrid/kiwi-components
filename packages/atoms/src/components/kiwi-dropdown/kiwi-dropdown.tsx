import {
  Component,
  ComponentInterface,
  Event,
  EventEmitter,
  h,
  JSX,
  Listen,
  Prop,
  State,
} from '@stencil/core';

@Component({
  tag: 'kiwi-dropdown',
  shadow: false,
})
export class KiwiDropdown implements ComponentInterface {
  private dropdownMenu: HTMLDivElement | undefined;

  /**
   * Css class to be applied to container.
   */
  @Prop()
  containerClass!: string;

  /** Type of the toggle button. */
  @Prop()
  toggleButtonType: 'default' | 'primary' | 'danger' | 'warning' | 'info' =
    'default';

  @State()
  private _open = false;

  private hover = true;

  /**
   * Event signaling this dropdown is being closed.
   */
  @Event()
  closeDropdown!: EventEmitter<void>;

  @Listen('click', { target: 'window' })
  handleOutsideClick(): void {
    if (!this.hover) {
      this.open = false;
    }
  }

  render(): JSX.Element {
    return (
      <div
        class={`btn-group dropdown ${this.open ? 'open' : ''} ${
          this.containerClass
        }`}
        onMouseEnter={this.handleMenuEnter}
        onMouseLeave={this.handleMenuLeave}
      >
        <button
          class={`btn btn-${this.toggleButtonType} dropdown-toggle`}
          onClick={this.handleDropdownToggle}
        >
          <slot name="dropdown-toggle" />
        </button>
        <div
          class="dropdown-menu"
          ref={(el) => (this.dropdownMenu = el as HTMLDivElement)}
        >
          <slot name="dropdown-content" />
        </div>
      </div>
    );
  }

  private handleDropdownToggle: () => void = () => {
    this.open = !this.open;
  };
  private handleMenuEnter: () => void = () => (this.hover = true);
  private handleMenuLeave: () => void = () => (this.hover = false);

  componentDidRender(): void {
    this.adjustPositionAtScreenEdge();
  }

  /**
   * Make sure that the dropdown does not stick over the right edge of the body
   */
  private adjustPositionAtScreenEdge(): void {
    if (this.dropdownMenu) {
      if (this.open) {
        const rect = this.dropdownMenu.getBoundingClientRect();

        if (rect.right > window.visualViewport.width) {
          const offset =
            rect.width > window.visualViewport.width
              ? Math.floor(rect.left)
              : Math.ceil(rect.right - window.visualViewport.width);

          this.dropdownMenu.style.translate = `-${offset}px`;
        }
      } else {
        this.dropdownMenu.style.translate = '';
      }
    }
  }

  private set open(isOpen: boolean) {
    this._open = isOpen;

    if (!isOpen) {
      this.closeDropdown.emit();
    }
  }

  private get open(): boolean {
    return this._open;
  }
}
