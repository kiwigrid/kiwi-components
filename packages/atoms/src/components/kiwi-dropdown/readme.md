# kiwi-dropdown



<!-- Auto Generated Below -->


## Properties

| Property              | Attribute                | Description                                                                                             | Type                                                        | Default     |
| --------------------- | ------------------------ | ------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------- | ----------- |
| `closeOnContentClick` | `close-on-content-click` | Setting this to true will cause the dropdown to close if a click is registered inside the dropdown-menu | `boolean`                                                   | `false`     |
| `containerClass`      | `container-class`        | Css class to be applied to container.                                                                   | `string`                                                    | `''`        |
| `toggleButtonType`    | `toggle-button-type`     | Type of the toggle button.                                                                              | `"danger" \| "default" \| "info" \| "primary" \| "warning"` | `'default'` |


## Events

| Event           | Description                                    | Type                |
| --------------- | ---------------------------------------------- | ------------------- |
| `closeDropdown` | Event signaling this dropdown is being closed. | `CustomEvent<void>` |


## Dependencies

### Used by

 - [kiwi-checkbox-dropdown-menu](../kiwi-checkbox-dropdown-menu)

### Graph
```mermaid
graph TD;
  kiwi-checkbox-dropdown-menu --> kiwi-dropdown
  style kiwi-dropdown fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
