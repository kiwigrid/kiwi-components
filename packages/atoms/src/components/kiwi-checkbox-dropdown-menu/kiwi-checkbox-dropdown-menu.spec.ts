import { newSpecPage } from '@stencil/core/testing';
import { KiwiDropdown } from '../kiwi-dropdown/kiwi-dropdown';
import { KiwiCheckboxDropdownMenu } from './kiwi-checkbox-dropdown-menu';

describe('kiwi-checkbox-dropdown-menu', () => {
  it('renders', async () => {
    const dropdown = await newSpecPage({
      components: [KiwiCheckboxDropdownMenu, KiwiDropdown],
      html: `
        <kiwi-checkbox-dropdown-menu>
          <span slot="dropdown-toggle">Toggle</span>
          <span slot="dropdown-content">Content</span>
        </kiwi-checkbox-dropdown-menu>
      `,
    });

    expect(dropdown.root).toMatchSnapshot();
  });
});
