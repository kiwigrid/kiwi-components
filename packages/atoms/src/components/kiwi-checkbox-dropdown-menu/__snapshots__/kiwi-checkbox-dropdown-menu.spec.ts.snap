// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-checkbox-dropdown-menu renders 1`] = `
<kiwi-checkbox-dropdown-menu>
  <!---->
  <kiwi-dropdown>
    <!---->
    <div class="btn-group checkbox-dropdown-menu dropdown">
      <button class="btn btn-default dropdown-toggle">
        <span slot="dropdown-toggle">
          Toggle
        </span>
      </button>
      <div class="dropdown-menu">
        <span slot="dropdown-content">
          Content
        </span>
      </div>
    </div>
  </kiwi-dropdown>
</kiwi-checkbox-dropdown-menu>
`;
